From 9e593244d3f9d6cb6fb6d2c0b5ebfea6c86d933b Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:14 -0400
Subject: [ethernet] qlcnic: Enhance semaphore lock access failure error message

Message-id: <1408626422-3941-91-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88615
O-Subject: [PATCH RHEL 7.1 090/138] qlcnic: Enhance semaphore lock access failure error message
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Harish Patil <harish.patil@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 1a51042bb8fc6317270bb70daf5eb1c57002340a

Signed-off-by: Harish Patil <harish.patil@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_hw.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_hw.c
index b94308c..ae88c2e 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_hw.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_hw.c
@@ -317,9 +317,7 @@ static void qlcnic_write_window_reg(u32 addr, void __iomem *bar0, u32 data)
 int
 qlcnic_pcie_sem_lock(struct qlcnic_adapter *adapter, int sem, u32 id_reg)
 {
- int timeout = 0;
- int err = 0;
- u32 done = 0;
+ int timeout = 0, err = 0, done = 0;
 
  while (!done) {
   done = QLCRD32(adapter, QLCNIC_PCIE_REG(PCIE_SEM_LOCK(sem)),
@@ -327,10 +325,20 @@ qlcnic_pcie_sem_lock(struct qlcnic_adapter *adapter, int sem, u32 id_reg)
   if (done == 1)
    break;
   if (++timeout >= QLCNIC_PCIE_SEM_TIMEOUT) {
-   dev_err(&adapter->pdev->dev,
-    "Failed to acquire sem=%d lock; holdby=%d\n",
-    sem,
-    id_reg ? QLCRD32(adapter, id_reg, &err) : -1);
+   if (id_reg) {
+    done = QLCRD32(adapter, id_reg, &err);
+    if (done != -1)
+     dev_err(&adapter->pdev->dev,
+      "Failed to acquire sem=%d lock held by=%d\n",
+      sem, done);
+    else
+     dev_err(&adapter->pdev->dev,
+      "Failed to acquire sem=%d lock",
+      sem);
+   } else {
+    dev_err(&adapter->pdev->dev,
+     "Failed to acquire sem=%d lock", sem);
+   }
    return -EIO;
   }
   msleep(1);
-- 
1.7.1